types.go

```diff
@@ -854,14 +854,20 @@ func (rr *NSEC) String() string {
 func (rr *NSEC) len(off int, compression map[string]struct{}) int {
 	l := rr.Hdr.len(off, compression)
 	l += domainNameLen(rr.NextDomain, off+l, compression, false)
-	lastwindow := uint32(2 ^ 32 + 1)
+	var lastwindow, lastlength uint16
 	for _, t := range rr.TypeBitMap {
 		window := t / 256
-		if uint32(window) != lastwindow {
-			l += 1 + 32
+		length := (t-window*256)/8 + 1
+		if window > lastwindow && lastlength != 0 { // New window, jump to the new offset
+			l += int(lastlength) + 2
+			lastlength = 0
 		}
-		lastwindow = uint32(window)
+		if window < lastwindow || length < lastlength {
+			continue // FIXME: packDataNsec would return Error{err: "nsec bits out of order"}
```

Comment:
What do you propose should be done here? `len` can’t and shouldn’t return an error, instead returning some sort of best effort length is appropriate as I believe this code already does.

Reply:
yeah, we can't error here, because we just want a length.

So the trigger is that the window count is out of order (has been a while, since this code was written)? I mean is the actual lenght calculation wrong, and the pack/unpack code a bit too liberal?

Reply:
I have little knowledge of NSEC RR. As far as I can tell this will return the actual length of the packet, valid or not, and packing will take care of erroring at a later stage. What really matters is that len() never return a length smaller than what could be used during packing, or we could overflow the buffer. I think it is fine to just drop the FIXME.